MultiDiskBundleStore

a10y opened the pull request:
Not ready for merge, but would appreciate feedback.

What's left right now:
- Remove disks

Commit:
Refactoring of BundleStore to make way for multiple disks

* Created a base class BaseBundleStore
* Moved BundleStore -> SingleDiskBundleStore
* Created a mock MultipleDiskBundleStore
* Added andreweduffy's fork of hash_ring to requirements

Commit:
Added beginnings of cl add-disk subcommand

Commit:
Allow users to specify which bundle_store to use, defaulting to SingleDiskBundleStore

Commit:
Make CodaLab assume SingleDiskBundleStore by default.

Commit:
Added soft_link command to path_util

Commit:
Implemented cl add-disk command with atomic semantics

Commit:
Added path_util.mkdtemp with contextmanager

Now, to create a temporary directory that will be deleted when a
context expires, you can do:

    with path_util.mkdtemp('testing') as tempdir:
        # logic in here ...
    # tempdir is deleted out here

Also updated the tests for SingleDiskBundleStore.

Commit:
Added safe rm-disk command

a10y commented:
@percyliang @kashizui @klopyrev This should be generally ready for review, only thing that would be left are a set of unit tests. Below I've written steps that I've been using to manually test:
1. Add the following line to your `~/codalab/config.json` somewhere not the last line:
   
   ```
   "bundle_store": "MultiDiskBundleStore",
   ```
2. Setup two "disks": `mkdir /tmp/my-first-disk /tmp/my-second-disk`
3. Add the first disk to your bundle store pool
   
   ```
   cl add-disk /tmp/my-first-disk
   ```
   
   This should create a link in `$HOME/.codalab/mdata/0` to `/tmp/my-first-disk`, which you should verify.
   In production that link would of course be to the mountpoint for a subdirectory on another drive, but the behavior is mostly the same (except for lack of atomic cross-disk renames, I think?).
4. Try uploading a bunch of bundles (like 6-7). Ensure that they are actually being placed on the phony disk:
   
   ```
   find -L $HOME/.codalab/
   ```
5. Add the second disk to the pool:
   
   ```
   cl add-disk /tmp/my-second-disk
   ```
   
   You should see a number of bundles say they are being relocated to the other disk. Check that some bundles (roughly half, yay hash rings!) are now underneath the `mdata/1` directory:
   
   ```
   find -L ~/.codalab/
   ```
6. Add a few more bundles, and convince yourself that they are being more-or-less evenly distributed across the "disks"
7. Try deleting one of the disks:
   
   ```
   cl rm-disk 1
   ```
   
   You should see that a number of bundles are "relocated", before the removal is successful. Verify that all bundles can still be `cl cat`'ed
8. Try removing the final disk, note that this should prevent you from doing this:
   
   ```
   cl rm disk 0
   ```

## Some Edge cases that I've tried to take into account:
- Deleting the last remaining disk is prevented
- If you remove a bunch of disks, when you start adding disks back it assigns the disk the lowest available number.

## Questions
- Current behavior when removing a disk is to not wipe bundles off of it, as it is being removed anyway. Is it better/more polite to delete the bundle contents off the disk, especially if it might be added back to the pool later?
- Should I squash some of these commits? It's a lot and a little messy at times
- I'm a little unsure what using a secondary temp directory gives us here. The usual rationale for using temp directories is because they allow you to copy all files to one location, and then do an atomic rename to the destination to simulate an atomic copy of all files. However, I don't know if this applies when we are copying through a symlink or across disks, possibly to multiple different disks. Note especially the `rm-disk` command, which needs to relocate bundles off of the dying disk to potentially several other disks. There's no simple way to do this atomically, but in the worse case we'd just be wasting space, as a bundle that exists on disk in two locations is only ever going to be referenced via one location (specifically the one that the hash ring dictates it should be in)

percyliang commented:
Thanks Andrew!  See comments below.

On Sat, Jan 16, 2016 at 1:16 AM, Andrew notifications@github.com wrote:

> @percyliang https://github.com/percyliang @kashizui
> https://github.com/kashizui @klopyrev https://github.com/klopyrev
> This should be generally ready for testing, only thing that would be left
> are a set of unit tests. Below I've written steps that I've been using to
> manually test:
> 
>    1.
> 
>    Add the following line to your ~/codalab/config.json somewhere not the
>    last line:
> 
>    "bundle_store": "MultiDiskBundleStore",
> 1. Setup two "disks": mkdir /tmp/my-first-disk /tmp/my-second-disk
>    3.
>    
>    Add the first disk to your bundle store pool
>    
>    cl add-disk /tmp/my-first-disk
>    
>    This should create a link in $HOME/.codalab/mdata/0 to
>    /tmp/my-first-disk, which you should verify.
>    In production that link would of course be to the mountpoint for a
>    subdirectory on another drive, but the behavior is mostly the same (except
>    for lack of atomic cross-disk renames, I think?).
>    4.
>    
>    Try uploading a bunch of bundles (like 6-7). Ensure that they are
>    actually being placed on the phony disk:
>    
>    find -L $HOME/.codalab/
>    
>    5.
>    
>    Add the second disk to the pool:
>    
>    cl add-disk /tmp/my-second-disk
>    
>    You should see a number of bundles say they are being relocated to the
>    other disk. Check that some bundles (roughly half, yay hash rings!) are now
>    underneath the mdata/1 directory:
>    
>    find -L ~/.codalab/
> 2. Add a few more bundles, and convince yourself that they are being
>    more-or-less evenly distributed across the "disks"
>    7.
>    
>    Try deleting one of the disks:
>    
>    cl rm-disk 1
>    
>    You should see that a number of bundles are "relocated", before the
>    removal is successful. Verify that all bundles can still be cl cat'ed
>    8.
>    
>    Try removing the final disk, note that this should prevent you from
>    doing this:
>    
>    cl rm disk 0
> 
> Some Edge cases that I've tried to take into account:
> - Deleting the last remaining disk is prevented
> - If you remove a bunch of disks, when you start adding disks back it
>   assigns the disk the lowest available number.
> 
> I think we should allow people to name the disks instead of numbering
> them.  So people would do `cl add-disk d2 /tmp/my-second-disk`.
> 
> Questions
> - Current behavior when removing a disk is to not wipe bundles off of
>   it, as it is being removed anyway. Is it better/more polite to delete the
>   bundle contents off the disk, especially if it might be added back to the
>   pool later?
> 
> We should delete the bundles off the old disk, of course only after we've
> verified that all the bundles have been copied successfully to the other
> disks.
> - Should I squash some of these commits? It's a lot and a little messy
>   at times
> 
> Up to you - your call.
> - I'm a little unsure what using a secondary temp directory gives us
>   here. The usual rationale for using temp directories is because they allow
>   you to copy all files to one location, and then do an atomic rename to the
>   destination to simulate an atomic copy of all files. However, I don't know
>   if this applies when we are copying through a symlink or across disks,
>   possibly to multiple different disks. Note especially the rm-disk
>   command, which needs to relocate bundles off of the dying disk to
>   potentially several other disks. There's no simple way to do this
>   atomically, but in the worse case we'd just be wasting space, as a bundle
>   that exists on disk in two locations is only ever going to be referenced
>   via one location (specifically the one that the hash ring dictates it
>   should be in)
> 
> I think each disk should have a data directory and a temp directory.  The
> principle is:
> 1) Moving from data to temp is instantaneous and atomic.
> 2) If operations fail, we can safely delete the temp disk and not affect
> the integrity of the system.

Suppose we have disks A1 A2 A3 and we're adding B.  So we need to move some
bundles from A\* to B.
1) copy A_.data to B.temp (takes a while)
2) move A_.data to A.temp, move B.temp to B.data (instantaneous), attach
disk B
3) delete A*.temp (might take a while)

If bad stuff happens before 2), we are left in the old state.  If bad stuff
happens after 2), we are in the new state.  This way, add-disk behaves
atomically modulo junk being left in *temp, which we can clear out.

We should document this logic in the code.

We need to probably test this quite a bit before checking into develop
(which gets periodically merged into master and deployed).

> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/308#issuecomment-172176126.

Commit:
Added an ls-disk command

a10y commented:
@percyliang This should be ready for you to look at. It's quite a lot but I rearranged the commits into something that hopefully is a more targeted set of changes.

I also factored `cleanup()` and `full_cleanup()` out into separate mixin classes. I was thinking that in the new pluggable BundleStore world, different stores can have a variety of behaviors based on a rich set of derived mixins, though it might not make sense with the way things are currently being done in which case I can simply take out the mixin.

klopyrev commented:
Three high-level comments:

This implementation doesn't work well with the multi-process REST design. In the multi-disk bundle store when you modify the ring object the version in only one of the processes would be modified. Should we reconsider using multiple processes, or can you work around it by saying requiring a server restart when you add a disk.

The new bundle store doesn't look thread-safe. Is the idea that no one would be accessing bundles when you add a disk?

Why is adding a disk done through the CLI, and not through the JSON config. That is to add a disk, you modify the JSON config and restart the server.

a10y commented:
@klopyrev 

> This implementation doesn't work well with the multi-process REST design. In the multi-disk bundle store when you modify the ring object the version in only one of the processes would be modified. Should we reconsider using multiple processes, or can you work around it by saying requiring a server restart when you add a disk.

Disks are only added/removed during downtime. If you look again, the ring is constructed from the filesystem. Only the administrator should have privileges to mess with the filesystem while the service is down. See below.

The ring object is constructed solely off of the contents of the file system, which will not be changing while the service is running.

> The new bundle store doesn't look thread-safe. Is the idea that no one would be accessing bundles when you add a disk?

Correct. An administrator will be adding/removing disks when the server has some sort of scheduled downtime.

> Why is adding a disk done through the CLI, and not through the JSON config. That is to add a disk, you modify the JSON config and restart the server.

I didn't make that choice but I'm assuming it's just to fit into the general UX of the rest of CodaLab, i.e. using the CLI for certain administrative tasks (setting permissions, running the bundle server and work manager, etc.). I agree that in the future a configuration-oriented approach would be the best way to go, as it lets you have your entire deployment state checked in and revisioned. However, I'm guessing that porting the CLI to be a wrapper around the REST API is at least a month or two off and this change is something that our existing servers could benefit from for load balancing.

kashizui commented on codalab/lib/bundle_store.py:
```diff
@@ -1,20 +1,109 @@
+        """
+        pass
+
+    def get_bundle_temp_location(self, identifer):
```
Is this still assuming that we are storing bundles on the files system? Could this be general enough to back a bundle store with something like S3? Or should `BundleStore`s only handle filesystem-based storage systems?

a10y replied:
Those are good questions. I feel like the get_bundle_location API is general enough to do something like ex. lookup in an S3 bucket, maybe the upload functionality could be factored out into a separate piece of functionally (ex. another mixin class called FileSystemUpload that allows uploading files to a bundle store.

In the current formulation though I think it's general enough, as upload_bundle doesn't assume anything about the storage format, all the options are for the bundle lookup done on disk. I feel like the API for uploads really could use improvement but it's not something I was looking to tackle in this PR.

percyliang commented:
On Wed, Feb 3, 2016 at 12:18 AM, Andrew notifications@github.com wrote:

> @klopyrev https://github.com/klopyrev
> 
> This implementation doesn't work well with the multi-process REST design.
> In the multi-disk bundle store when you modify the ring object the version
> in only one of the processes would be modified. Should we reconsider using
> multiple processes, or can you work around it by saying requiring a server
> restart when you add a disk.
> 
> Disks are only added/removed during downtime. If you look again, the ring
> is constructed from the filesystem. Only the administrator should have
> privileges to mess with the filesystem while the service is down. See below.
> 
> The ring object is constructed solely off of the contents of the file
> system, which will not be changing while the service is running.
> 
> The new bundle store doesn't look thread-safe. Is the idea that no one
> would be accessing bundles when you add a disk?
> 
> Correct. An administrator will be adding/removing disks when the server
> has some sort of scheduled downtime.
> 
> Why is adding a disk done through the CLI, and not through the JSON
> config. That is to add a disk, you modify the JSON config and restart the
> server.
> 
> I didn't make that choice but I'm assuming it's just to fit into the
> general UX of the rest of CodaLab, i.e. using the CLI for certain
> administrative tasks (setting permissions, running the bundle server and
> work manager, etc.). I agree that in the future a configuration-oriented
> approach would be the best way to go, as it lets you have your entire
> deployment state checked in and revisioned. However, I'm guessing that
> porting the CLI to be a wrapper around the REST API is at least a month or
> two off and this change is something that our existing servers could
> benefit from for load balancing.
> 
> Yes, the idea is that this command would be run completely off-line in
> single-threaded mode when nothing else is running.  No need to deal with
> REST.

I think adding/removing disks is a pretty delicate issue that we want to
have people manually do with 'cl add-disk' and 'cl remove-disk', and not
leave to some automatic mechanism.

This is a pretty important feature to get right (so we don't lose people's
data!), so it'd be good if @kashizui could take a careful look at this too.

> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/308#issuecomment-179082583.

percyliang commented:
Some comments about usage:
- Let's rename add-disk, rm-disk, ls-disk to disk-add, disk-rm, disk-ls so they  don't interfere with the usual rm, ls commands, and you can see all the disk  related commands using autocomplete.
- Let's reverse the arguments to `cl disk-add <disk-name> <path>`, which is more natural.  Add an example usage: 'cl disk-add disk1 /mnt/codalab1'
- Help message for `disk-rm` is out of date.  `disknum` should be <disk-name>, and  people should use `disk-ls` to find the disk names.
- The current directory names are a bit confusing: `.codalab/mdata` currently  points to a directory that has another mdata and a mtemp directory in it.  I  think the `.codalab/mdata` should be renamed `.codalab/disks` and  `.codalab/mtemp` should just be `.codalab/temp`.
- Let's make the migration from SingleDiskStore to MultiDiskStore more  seamless.  I think we don't need a 'bundle_store' in the config file.  The  default should just MultiDiskStore.  And if no .codalab/mdata directory  exists, then one called 'default' should just be created and link to  .codalab/data.  Then people don't have to do any additional work to just start  using the new system, and this is backwards compatible with SingleDiskStore  still.

@andreweduffy

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -1,20 +1,109 @@
+        self = args[0]
+        if self.ring.get_node("DOESN'T MATTER") is None:
+            print >> sys.stderr,"""
+Error: No disks available.
```
If this happens, automatically create `default` disk (see overall comments).

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -169,41 +259,413 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+    Q: What happens when the disk a bundle is supposed to go to runs out of space?
+    A: Need to avoid this scenario by proactively adding new disk mounts when the time arises, and then
+       redistributing bundles when they need to be redistributed.
+    """
```
Write a comment here that if we have disks D1 and D2 with sizes 3N and N, then one should add 3 logical disks under D1.  Basically users have to manage their own disks.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -169,41 +259,413 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+    """
+
+    # Location where MultiDiskBundleStore keeps production and temp data relative to each disk
+    DATA_SUBDIRECTORY = 'mdata'
```
rename to 'disks'

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -169,41 +259,413 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+
+    # Location where MultiDiskBundleStore keeps production and temp data relative to each disk
+    DATA_SUBDIRECTORY = 'mdata'
+    TEMP_SUBDIRECTORY = 'mtemp'
```
rename to 'temp'

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -169,41 +259,413 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+        keys to precomputed statistics about the new data directory.
+        """
+        to_delete = []
+
```
Should figure out what disk to use first and then upload everything to the 'temp' directory of **that** disk.  That way, moving to the data directory is cheap and we can also fail early if that disk is broken or something.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -169,41 +259,413 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+            data_hash = '0x%s' % (path_util.hash_directory(temp_path, dirs_and_files),)
+
+            # Determine which disk this will go on
+            disk_choice = self.ring.get_node(data_hash)
```
I see the problem, the disk choice depends on the data_hash, not the UUID.  That's really too bad.  This provides another incentive to store bundles by uuid rather than data_hash.  Let's not do this in this PR, but add an another issue for this.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -169,41 +259,413 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+        try:
+            path_util.make_directory(mtemp)
+        except:
+            print >> sys.stderr, "Could not make directory %s on disk %s, aborting" % (MultiDiskBundleStore.TEMP_SUBDIRECTORY, target)
```
Print out the actual mtemp path.

percyliang commented on requirements.txt:
```diff
@@ -6,3 +6,5 @@ watchdog==0.8.3
 # https://github.com/kislyuk/argcomplete/pull/118
 git+https://github.com/kashizui/argcomplete.git@allow-overridden-subparsersaction
+# Use agd@stanford.edu's fork of hash_ring with extra features
+git+https://github.com/andreweduffy/hash_ring.git@master
```
This is kind of unfortunate...what features do we need?
Can we just put the code directly into CodaLab?
I don't think it's good having tons of dependencies if they're going to be custom anyway.

a10y replied:
There are no good consistent hashing libraries for Python :/ (see https://github.com/Doist/hash_ring/issues/11)

I added the ability and dynamically add and remove nodes in the hash ring, which doesn't exist in the original version. If you'd like I can move the changes over into CodaLab, but the original repo that I forked from doesn't have any license, which apparently means that "default copyright rules apply" i.e. you can't use their code directly https://help.github.com/articles/open-source-licensing/#what-happens-if-i-dont-choose-a-license. I doubt that's what the author intended, but just thought you should be aware.

I can give the ownership of the repo to CodaLab, and then we can maintain our own fork if you think that'd be better.

a10y replied:
Actually the setup.py file apparently states that the project is under the BSD license (though no explicit license is included with the source), so that means that we can reproduce and use it. I can definitely make this a CodaLab-official maintained fork if you think that'd be best. I can also go ahead and write one from scratch based off of another implementation, let me know what you think.

percyliang commented on tests/lib/bundle_store_test.py:
```diff
@@ -86,8 +96,16 @@ def hash_directory(path, dirs_and_files=None):
+    """
+    Ensure that adding/removing disks works as expected
+    """
+    pass
```
We should add some basic tests before this PR gets checked in.

percyliang commented:
I wonder if the name 'disk' is confusing, because it's a logical disk, not a physical disk, and people are expected to actually create several 'disks' on an actualy physical disk.

Maybe we should rename things to 'partition'?  Then we would have:

```
cl bundle-store-add-partition
cl bundle-store-rm-partition
cl bundle-store-ls-partitions
```

And `.codalab/partitions`, `.codalab/temp`, and in each partition, we have directories `data` and `temp`.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -1,20 +1,109 @@
+
+
+
+class SingleDiskBundleStore(BaseBundleStore, BundleStoreCleanupMixin):
```
We should be able to delete this entirely.

Commit:
Merged in UUID-based storage

Commit:
Leftover fixes from the merge

Commit:
Fixes for cleanup

Commit:
thigns

Commit:
Merge branch 'develop' into multi-disk

a10y commented:
@percyliang 

I agree that perhaps partition (or "shard"?) might be better naming. I feel like `bundle-store-add-partition` is might be a bit too much to type, though I do realize people won't be using these commands too often.

percyliang commented:
I think partition is better.  Shard suggests something a lot smaller.

How about bs-add-partition?

On Wed, Feb 24, 2016 at 2:26 PM, Andrew notifications@github.com wrote:

> @percyliang https://github.com/percyliang
> 
> I agree that perhaps partition (or "shard"?) might be better naming. I
> feel like bundle-store-add-partition is might be a bit too much to type,
> though I do realize people won't be using these commands too often.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/308#issuecomment-188482937.

Commit:
Rename disks -> partitions

Commit:
Fix for cl rm --data-only

If we do a data-only rm, we delete the data_hash as well as the
raw data from the BundleStore, leaving the metadata in the
database. We always delete the hash first, so make sure that
files which are old don't just get randomly deleted.

Commit:
Added hash_ring repository into the repo.

Taken from andreweduffy/hash_ring and placed inside of here.
The original version of hash_ring did not have an explicit
license in the repository, meaning we may need to take extra
steps to legitimize this in the future, but for now I don't
think anyone will protest to this being shipped with CodaLab.

Commit:
Make CircleCI tests run test-cli with MultiDiskBundleStore

Added ci-multidisk script to configure MultiDisk, added to circle.yml

Commit:
Ensure that bs-ls-partitions only works for MultiDiskBundleStore

Commit:
Updated help message

Commit:
Added migration script

a10y commented:
@percyliang 

That should wrap it up for this feature. I also went ahead and added a migration script, which basically performs the following steps:
1. Configuration file is updated to have `"bundle_store": "MultiDiskBundleStore"`
2. Sets up a new initial partition for MultiDisk. User can provide a directory to setup their new base partition, or just go with `$CODALAB_HOME/dummy`. Name is `default`
3. `$CODALAB_HOME/partitions/default/bundles` is replaced with `$CODALAB_HOME/bundles` from the single disk days.

percyliang commented:
@andreweduffy: I don't think we need the explicit migration.  Some more minor comments below.

1) MultiDiskBundleStore can just be the default without an explicit need for modifying config.json.  Let's deprecate SingleDiskBundleStore.

2) If 'partitions' doesn't exist, then just create a partition called 'default' that points to $CODALAB_HOME.  This keeps backward compatibility, but is also a reasonable default.  Don't call it 'dummy'.

3) Why are relative paths not allowed when you add-partition?  I would allow them but just resolve to the absolute path when you make the symlink.

4) Could we swap the order of arguments for bs-add-partition ('name path' rather than 'target name') - don't call it target, because target refers to uuid/subpath in CodaLab.

Commit:
Swap order of bs-add-partition arguments

a10y commented:
@percyliang In that case do you want me to delete SingleDiskBundleStore completely?

Commit:
Allow relative paths in bs-add-partition

Commit:
MultiDiskBundleStore creates a default partition that points to CODALAB_HOME

This makes transitioning from the old SingleDiskBundleStore over
to MultiDiskBundleStore zero-effort. The CODALAB_HOME/partitions
directory is created, and then CODALAB_HOME/partitions/default
is setup to link to CODALAB_HOME, which means that it will use
the existing CODALAB_HOME/bundles and CODALAB_HOME/temp as the
bundle and temp directories.

percyliang commented:
Yes - we should be able to simulate the same behavior with multi disk.
On Feb 26, 2016 2:09 PM, "Andrew" notifications@github.com wrote:

> @percyliang https://github.com/percyliang In that case do you want me
> to delete SingleDiskBundleStore completely?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/308#issuecomment-189504338.

Commit:
Removed SingleDiskBundleStore

MultiDiskBundleStore by default will simulate behavior of
SingleDisk, so it has been officially deprecated. Updates all
tests that referenced SingleDiskBundleStore and replaced to test
MultiDiskBundleStore.

Commit:
Changed MISC_TEMP_SUBDIRECTORY to prevent collisions

CodaLabManager sets the tempfile path to
MultiDiskBundleStore.MISC_TEMP_SUBDIRECTORY, which is then used
by BundleRPCServer. We want to avoid mixing things into the
temp_dir, as it's better to assume the temp directory is in
somewhat clean order.

a10y commented:
@percyliang I think that should integrate all of your comments. A couple of things:
- I went ahead and changed the "bundle store temp dir" to be different from `CODALAB_HOME/temp` so that way temp files from BundleRPCServer end up inside of the default partitions temp directory.
- I've tested the usage and setup of default partition both with and without a pre-existing SingleDiskBundleStore and it seems okay.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -161,28 +226,188 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+                correct_partition = self.ring.get_node(bundle)
+                if correct_partition != partition:
+                    # Reposition the node to the correct partition
+                    print >> sys.stderr, "Marking %s for relocation" % bundle
```
print out something more precise: copying from_path to to_path

a10y replied:
sounds good

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -161,28 +226,188 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
-        if not dry_run:
-            path_util.remove(absolute_path)
+        try:
+            bundle = model.get_bundle(uuid)
```
Don't need to pass in / query the model anymore...since the uuid just determines the location.
Also, seems like except_bundle_uuids isn't used, so we can get rid of that.

a10y replied:
I thought that when we do a full_cleanup, we need to look for items that are old and also not in the DB anymore? Or items that are in the DB but no longer have a data_hash (because they were previously deleted using `rm --data-only`.

a10y replied:
Though now that I think about it, this would also go ahead and delete bundles that are still running and so their data_hash is also None, so I also need to check bundle state and make sure it's not marked as running?

percyliang replied:
I think full_cleanup should actually get removed - it's never been used before.  It was to go through and delete old temp files, but your other bundle store health-checking script should be doing that.  Maybe we should have that as a cl command - `cl bs-check`.

If a bundle is running, the CLI won't let you delete the bundle, unless you pass in `--force`.  In general, if possible, the bundle store shouldn't depend on the model.

percyliang commented on codalab/lib/hash_ring/ring.py:
```diff
@@ -0,0 +1,178 @@
+
+    :copyright: 2008 by Amir Salihefendic.
+    :license: BSD
+    Modifications by Andrew Duffy <root@aduffy.org>
```
Summarize modifications briefly?

a10y replied:
will do

percyliang commented on scripts/ci-multidisk.sh:
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
```
1) Let's put the test in test-cli.py (we need to of course do it in a temporary CODALAB_HOME - see the 'copy' test).

2) Let's avoid having the CONFIG copied here (because then we have to keep it updated).  In the test, the config can be automatically generated by `cl`.

a10y replied:
This isn't actually a test, it was just setting up the bundle store before we ran test-cli. This predates the removal of SingleDisk before Multi was the default, so if we go ahead and delete this script we should be fine.

percyliang commented:
@andreweduffy LGTM modulo comments

a10y commented:
@percyliang Replied to your comments, the cosmetic ones I've fixed but I'm not certain about cleanup/full_cleanup.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -161,28 +226,188 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+            if not dry_run:
+                path_util.remove(absolute_path)
 
     def full_cleanup(self, model, dry_run):
```
Let's delete this (and replace with your general bundle store health checking script).

a10y replied:
@percyliang This is what's called when you run `cl rm` though, are we just going to leave things in the bundle store by default and then remove them later?

percyliang replied:
@andreweduffy no cl rm just does cleanup, not full_cleanup.  Let's remove the cleanup command (which calls full_cleanup) from the CLI.

a10y replied:
Oh whoops misread the location of the comment. I'll go ahead and restructure things, removing `full_cleanup` and the dependency on the model, as well as the cleanup command. The BundleStore health check command will go into a separate PR once it's done.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -1,65 +1,127 @@
+class BundleStoreCleanupMixin(object):
+    """A mixin for BundleStores that wish to support a cleanup and full cleanup operation.
+    """
+    def cleanup(self, model, uuid, except_bundle_uuids, dry_run):
```
These should be updated too.

Commit:
Cosmetic fixes + removal of ci-multidisk script

Commit:
Removed full_cleanup, model from MultiDiskBundleStore

The full cleanup command is going away, it will be replaced with
a general bundle store health check command in the future. Also
removed dependence on the model in cleanup(), as well as removes
the unused relevant_uuids parameter that was leftover from
deduplicated storage.

Commit:
Fixed indentation of bundle store commands
